src/App.elm

```diff
@@ -71,133 +76,176 @@ update msg model =
-            "Malformed url: " ++ url
+        definitions =
+            List.map viewDefinitionListing content.definitions
+    in
+    div [] (namespaces ++ definitions)
 
 
-viewNamespaceChild : Namespace.NamespaceChild -> Html Msg
-viewNamespaceChild namespaceChild =
-    case namespaceChild of
-        Namespace.SubNamespace name size ->
-            div [] [ text name ]
+viewNamespaceListingContent : HashSet -> WebData NamespaceListingContent -> Html Msg
+viewNamespaceListingContent expandedNamespaceListings content =
+    case content of
+        Success loadedContent ->
+            viewLoadedNamespaceListingContent expandedNamespaceListings loadedContent
 
-        Namespace.Type name hash ->
-            div [] [ text name ]
+        Failure err ->
+            text (Api.errorToString err)
 
+        NotAsked ->
+            UI.nothing
 
-viewNamespaceTree : RemoteData.WebData Namespace.Namespace -> Html Msg
-viewNamespaceTree namespace =
-    let
-        content =
-            case namespace of
-                RemoteData.Success ns ->
-                    List.map viewNamespaceChild ns.children
+        Loading ->
+            UI.spinner
 
-                RemoteData.Failure err ->
-                    [ text (errorToString err) ]
 
-                RemoteData.NotAsked ->
-                    [ text "Loading" ]
+viewNamespaceListing : HashSet -> NamespaceListing -> Html Msg
+viewNamespaceListing expandedNamespaceListings (NamespaceListing hash fqn content) =
+    let
+        namespaceContent =
+            if HashSet.member hash expandedNamespaceListings then
+                div [ class "namespace-content" ]
+                    [ viewNamespaceListingContent
+                        expandedNamespaceListings
+                        content
+                    ]
 
-                RemoteData.Loading ->
-                    [ text "Loading" ]
+            else
+                UI.nothing
     in
-    div [ class "namespace-tree" ] content
+    div []
+        [ a
+            [ class "node namespace"
+            , onClick (ToggleExpandedNamespaceListing hash)
+            ]
+            [ text (unqualifiedName fqn) ]
+        , namespaceContent
+        ]
 
 
-viewNamespacePath : Namespace.Path -> Html msg
-viewNamespacePath (Namespace.Path path) =
+viewAllNamespaces : HashSet -> WebData NamespaceListing -> Html Msg
+viewAllNamespaces expandedNamespaceListings namespaceRoot =
     let
-        namespaceLinks =
-            path
-                |> List.Nonempty.map (\p -> a [ class "namespace" ] [ text p ])
-                |> List.Nonempty.toList
-
-        slash =
-            span [ class "slash" ] [ text "/" ]
-    in
-    label
-        [ class "definition-namespace-path" ]
-        (List.intersperse slash namespaceLinks)
+        listings =
+            case namespaceRoot of
+                Success (NamespaceListing _ fqn content) ->
+                    viewNamespaceListingContent expandedNamespaceListings content
```

Comment:
The root should **not** have a node on its own so we render the children directly instead